protocol/superchainerc20/L1-native-factory.md

```diff
@@ -0,0 +1,307 @@
+
+After the first deployment, it is possible to deploy from L1 or from the factory that already deployed and stored the metadata using interop.
+
+An alternative design to deploy from L1 is to use the `L1 blockhash` along with storage proofs to prove the properties. This would allow to initialize deployments from L2 as well (the factory would need to implement two external functions in this case).
+
+The factory will store
+
+```solidity
+
+mapping(address => DeploymentData) public deploymentsData;
+```
+
+that maps each L1 address to the `DeploymentData` struct
+
+```solidity
+struct DeploymentData {
+  address superchainERC20;
+  uint8 decimals;
+  string name;
+  string symbol;
+}
+```
+
+## Implementation
+
+**Proxy Factory**
+
+Will follow the [`Proxy.sol` implementation](https://github.com/ethereum-optimism/optimism/blob/v1.1.4/packages/contracts-bedrock/src/universal/Proxy.sol) and `delegatecall()` to the factory implementation address.
+
+**Factory Implementation**
+
+```solidity
+contract SuperchainERC20Factory is Semver {
+  address public constant NATIVE_INITIALIZER = 0x0...;
+
+  // address of superchainERC20 is packed with decimals to save a slot
+  // alternative order would be to have superchainERC20 at the end, and follow the name, symbol, decimal order
+  // of ERC20
+  struct DeploymentData {
+    address superchainERC20;
+    uint8 decimals;
+    string name;
+    string symbol;
+  }
+
+  mapping(address => DeploymentData) public deploymentsData;
+
+  event SuperchainERC20Deployed(address indexed superchainERC20, address indexed remoteToken, uint8 decimals, string name, string symbol);
+
+  constructor() Semver(1, 0, 0) {}
+
+  // This function can be called from L1 or L2. A real implementation would probably move the
+  // deployment logic to an internal function and split deploy() into deployFromL1() and deployFromL2()
+  // external functions to minimize branching.
+  function deploy(address _remoteToken, uint8 _decimals, string memory _name, string memory _symbol) external returns (address _superchainERC20) {
+    if (msg.sender == Predeploys.L2CrossChainMessenger) {
+        require(IL2CrossChainMessenger(Predeploys.L2CrossChainMessenger).xDomainMessageSender() == NATIVE_INITIALIZER, "Invalid initializer");
+    } else if (msg.sender == Predeploys.L2ToL2CrossDomainMessenger) {
+        require(IL2toL2CrossDomainMessenger(Predeploys.L2ToL2CrossDomainMessenger).crossDomainMessageSender() == address(this), "Invalid cross-domain sender");
+    } else {
+        revert("Unauthorized caller");
+    }
+
+    require(deployments[_remoteToken] == address(0), "Token already deployed");
```

Comment:
Do we strictly want this property? Factory will only deploy 1 SuperchainERC20 token per L1 token? If so, why?